File: S/1260-shift-2d-grid/main.go
package main

import (
	"fmt"
)

func shiftGrid(grid [][]int, k int) [][]int {
	if len(grid) == 0 {
		return grid
	}

	row := len(grid)
	col := len(grid[0])

	rowShift := k / col
	colShift := k % col

	ret := make([][]int, row)
	for i := 0; i < row; i++ {
		ret[i] = make([]int, col)
	}
	for i := 0; i < row; i++ {
		for j := 0; j < col; j++ {
			r := (i + rowShift)
			c := (j + colShift)
			if c >= col {
				c %= col
				r++
			}
			r %= row
			ret[r][c] = grid[i][j]
		}
	}
	return ret
}

func main() {
	fmt.Println(shiftGrid([][]int{{1, 2, 3}, {4, 5, 6}, {7, 8, 9}}, 1))

}

File: S/1260-shift-2d-grid/main.rs
impl Solution {
  pub fn shift_grid(grid: Vec<Vec<i32>>, k: i32) -> Vec<Vec<i32>> {
    let (row, col) = (grid.len(), grid[0].len());
    let mut ans: Vec<Vec<i32>> = vec![vec![0; col]; row];
    let k = k as usize % (row * col);
    for r in 0..row {
      for c in 0..col {
        let idx = (r * col + c + k) % (row * col);
        ans[idx / col][idx % col] = grid[r][c];
      }
    }
    ans
  }
}
